File: ReverseList2.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Thu Dec 19 10:32:15 2019

@author: tanvirkaur
"""
# iterative solution
# time complexity = O(n)
# space complexity = O(1) if we donot consider the recursive stack

class Solution:
    def reverseList(self, head: ListNode) -> ListNode:
        # base case
        if head == None or head.next == None:
            return head
        r = self.reverseList(head.next)
        head.next.next = head
        head.next = None
        return r

File: reverselist2.java
// Time Complexity : O(n)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode :yes
// Any problem you faced while coding this : no


public class reverselist2 {
    /**
     * Definition for singly-linked list.
     * public class ListNode {
     *     int val;
     *     ListNode next;
     *     ListNode() {}
     *     ListNode(int val) { this.val = val; }
     *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
     * }
     */
        public ListNode reverseList(ListNode head) {
            ListNode prev = null;
            ListNode curr = head;

            while(curr != null){
                ListNode next = curr.next;
                curr.next = prev;
                prev = curr;
                curr = next;
            }

            return prev;

        }
    }
